### PR TITLE
Integrate Prometheus Metrics and Enhance Observability for WebSocket and WebRTC Connections

### PR DESCRIPTION
**Pull request by jhmin99**
### Description
This pull request integrates Prometheus metrics collection and monitoring into the project to enhance observability. Metrics are captured for `WebSocket connections`, `WebRTC connections`, `CPU usage`, `memory usage`, and `network usage`. These changes lay the foundation for improved monitoring and debugging capabilities. The controller now tracks connection metrics and updates them during various operations, while Prometheus is configured to scrape and visualize these metrics.

### Changes Made
- Metrics System: Introduced a Prometheus-based metrics system with custom metrics tracking:
   - WebSocket connections (websocket_connections_total)
   - WebRTC connections (webrtc_connections_total)
   - CPU usage (cpu_usage_percentage)
   - Memory usage (memory_usage_bytes)
   - Network usage (network_usage_bytes with inbound/outbound labels)
- Prometheus Integration: Added Prometheus configuration with scrape jobs.
- Grafana Integration: 
   - Pre-configured Grafana dashboard added:
      - [Gopring.json](https://github.com/user-attachments/files/18017634/Gopring.json)
   - Dashboard visualizes all metrics with gauges and labels (e.g., direction for network usage). 
- Controller Updates:
   - Integrated metrics updates into SocketController.
   - Incremented/decremented WebSocket and WebRTC connection counts in appropriate lifecycle events.
- Docker Setup: Configured Prometheus and Grafana services in docker-compose.yml.

### Checklist
- [X] Tests have been added/updated or not required
- [X] didn't break anything

### Additional Context
- The JSON file [Gopring.json](https://github.com/user-attachments/files/18017634/Gopring.json) can be imported into Grafana for immediate setup.
- Instructions for importing the dashboard and configuring Prometheus are included in the updated README.md.
- Screenshots of the dashboard are included for reference:

스크린샷 2024-12-05 오후 1 59 38

### How to Use the Grafana Dashboard

1. Import the attached JSON file: 
   - Go to Dashboards > Import in Grafana. [Gopring.json](https://github.com/user-attachments/files/18017634/Gopring.json)
   - Paste the JSON content or upload the file.
2. Connect the dashboard to your Prometheus data source.
3. Verify that metrics are displayed correctly in the panel.

### CODE REVIEW

File: metric/metric.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// Start initializes and starts the metrics HTTP server.
+func (m *Metrics) Start(stop <-chan struct{}) {
+	m.httpServer = &http.Server{
+		Addr:              fmt.Sprintf(":%d", m.config.Port),
+		Handler:           promhttp.Handler(),
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
```
Comment (window9u) [inline metric/metric.go:84]: Consider subscribing to the broker's events and publishing metrics as demonstrated below:

[Reference Code](https://github.com/Gopring/PDN-WebRTC/blob/main/media/media.go#L45)

### Example: Subscribe to Events and Publish Metrics

```go
// Subscribe to broker events
pushEvent := m.broker.Subscribe(broker.ClientMessage, broker.PUSH)
// (Note: The `connectedEvent` subscription is available on the coordinator branch and will be published soon)
// connectedEvent := c.broker.Subscribe(broker.Media, broker.CONNECTED)

for {
    var err error
    select {
    case event := <-pushEvent.Receive():
        // Example: Publish a metric for an increase in push requests
        // Metric.Publish(IncreasePushRequest)

    case event := <-connectedEvent.Receive():
        // Example: Publish a metric for increased server connections
        // Metric.Publish(IncreaseServerConnection)
    }

    // Handle any errors during event processing
    if err != nil {
        log.Printf("Failed to handle event in Media: %v", err)
    }
}
```

File: metric/metric.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// Default values for metrics configuration.
+const (
+	DefaultMetricsPort = 9090
+	DefaultMetricsPath = "/metrics"
+)
+
+// New creates a new Metrics instance with the specified configuration.
+func New(config Config) *Metrics {
```
Comment (window9u) [inline metric/metric.go:42]: Consider using dependency injection for the `Broker` in `Metric`.

File: signal/handler/socket.go [hunk 3]
```diff
@@ -32,7 +35,10 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		return
 	}
+
+	h.metrics.DecrementWebSocketConnections()
 	defer func(conn *websocket.Conn) {
+		h.metrics.DecrementWebSocketConnections()
```
Comment (window9u) [inline signal/handler/socket.go:41]: How do you think about publish event that socket disconnection, and subscribe it in Metric?

**Comment by window9u** [comment]
LGTM!!

**Comment by jhmin99** [comment]
Update to [gopring.json](https://github.com/user-attachments/files/18060751/gopring.json)

**Pull request merged**